### DISCUSSION TITLE
Step Inputs / Cache

### DISCUSSION THREAD

**Discussion by rawkode** [juspay/omnix#247] [Q&A]
Very interesting project. I'm still working my way through the examples in this repository and the code, but something I was curious about is restricting steps based on a subset of the directory tree.

Assuming I wanted a step to run `npm install`, it's only input would be `package.json` and it's output would be `node_modules`

If I was to run Omnix in a monorepository, and say use `bun install`, I'd only want this to run again if a `package.json` nested within the repository has changed.

Is this something Omnix will handle or have I misunderstood the goals?

**Comment by srid** [comment]
@rawkode  [￼`om ci`￼](https://omnix.page/om/ci.html) is all about running CI builds for your flake(s). The use-case you are describing seems orthogonal to `om ci` (including the various CI [‘steps’](https://omnix.page/om/ci.html#custom)).

> If I was to run Omnix in a monorepository, and say use bun install, I'd only want this to run again if a package.json nested within the repository has changed.

Here `bun install` seems to be the JavaScript equivalent of Rust’s `cargo install`. Omnix or not - generally, when Nixifying a project, we don’t run these commands except through Nix derivations (packages or checks), such as for example https://github.com/NixOS/npm2nix which runs `npm install`. You can run custom commands via custom steps, per the link above, but the repository packages (including dependency fetching) should normally be built as Nix derivations.

What you want sounds more like a `Makefile` tool that will track dependencies.
